Feature branch: Plugins: make "browse" the default route and improve behavior

sirbrillig opened the pull request:
This is a work-in-progress PR to update the Plugins page so that browsing plugins happens at `/plugins` (previously `/plugins/browse`, although that will still work) and managing plugins happens at `/plugins/manage` (previously `/plugins`).

There are several other improvements as part of this effort, and this PR may end up becoming multiple PRs as it progresses.

Here's a brief list of the changes this project is meant to address:

- [x] Make /plugins go to browse page
- [x] Make sure browse page still supports filter 
- [x] Make sure browse page still supports category 
- [x] Make sure browse page still supports search
- [x] Make sure browse page still supports multi-site
- [x] Move manage page to /plugins/manage
- [x] Add suggested search tabs to browse page
- [x] Add "Manage" button to Browse page header
- [x] Add "Manage" button to sidebar
- [x] Make sure manage page still supports filters
- [x] Make sure manage page still supports search
- [x] Change manage page to 3 columns
- [x] Add featured and popular plugins below installed on Manage page
- [x] Add "add plugin" button to Manage page header
- [x] Make sure bulk management still works
- [ ] Redirect /plugins/browse to /plugins, keeping filter, site, and search
- [ ] Add Jetpack features as "plugin" blocks in browse page
- [ ] Add error message notice inside plugin item; click error notice to show error
- [ ] Move "Upload Plugin" button in Manage page to header

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +195,107 @@ const PluginsBrowser = React.createClass( {
+		// TODO: The Search does not get the search text when a suggested term is clicked
+		return (
+			<SectionNav
+				selectedText={ this.props.translate( 'Suggested Searches', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +195,107 @@ const PluginsBrowser = React.createClass( {
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Appearance', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Writing', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **49** times:
`translate( 'Writing', { context: 'settings screen'} )` *ES Score: 17*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +195,107 @@ const PluginsBrowser = React.createClass( {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Appearance', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Appearance', { context: 'Filter label for plugins list'} )` *ES Score: 19*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +195,107 @@ const PluginsBrowser = React.createClass( {
+		const site = this.props.site ? '/' + this.props.site : '';
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'Security', { context: 'settings screen'} )` *ES Score: 17*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +195,107 @@ const PluginsBrowser = React.createClass( {
+	getSearchBar() {
+		const site = this.props.site ? '/' + this.props.site : '';
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Engagement' )` *ES Score: 26*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +203,113 @@ const PluginsBrowser = React.createClass( {
+
+		return (
+			<SectionNav
+				selectedText={ this.props.translate( 'Suggested Searches', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +203,113 @@ const PluginsBrowser = React.createClass( {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Appearance', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Appearance', { context: 'Filter label for plugins list'} )` *ES Score: 19*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +203,113 @@ const PluginsBrowser = React.createClass( {
+	getSearchBar() {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'Security', { context: 'settings screen'} )` *ES Score: 17*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +203,113 @@ const PluginsBrowser = React.createClass( {
+
+	getSearchBar() {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Engagement' )` *ES Score: 26*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

sirbrillig commented:
@jsnajdr I rebased the branch to help get rid of the infinite loop errors we were seeing.

Using `WrappedSearch` to re-create the `Search` component and get it to use `initialValue` again was very clever, but it's quite an unusual trick and I worry that it could cause a lot of confusion later. 

We still might want to go that route, but just to experiment, I pushed a few commits to use local state to update `Search` directly instead. Let me know what you think! Your version is still there so we can go back if we need to.

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +199,111 @@ const PluginsBrowser = React.createClass( {
 
+		return (
+			<SectionNav
+				selectedText={ this.props.translate( 'Suggested Searches', {
```
:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

jsnajdr commented:
> We still might want to go that route, but just to experiment, I pushed a few commits to use local state to update `Search` directly instead.

Hello @sirbrillig! Unfortunately, I can break it just by typing quickly :) The debounced `onSearch` triggers this asynchronous funny dance:
![search-dance](https://user-images.githubusercontent.com/664258/29739447-41709494-8a3e-11e7-946c-41f4c6d8dc65.gif)

> Let me know what you think! Your version is still there so we can go back if we need to.

I'm afraid the `WrapperSearch` component is the best solution for now if we want to move forward with the plugins project. The `Search` component fights back very vigorously against any attempt to change it :)

jsnajdr commented:
@sirbrillig I reverted your two commits that were attempting to use state to control the `Search` component. The HOC version with `WrappedSearch` seems to work better. I created and pushed a branch named `try/plugin-search-controlled` which still has these commits, if you want to keep them and work on them further. If you don't need them any more, you can just delete the branch.

jsnajdr commented:
In the latest few commits, I implemented a first version of three-column view for Plugin Management. This is how it looks:

<img width="1052" alt="plugins-manage-3-columns" src="https://user-images.githubusercontent.com/664258/29774928-9641de08-8c03-11e7-8ecb-b6593767570e.png">

It's still very imperfect and will require a lot of further CSS and markup tuning. For example, the layout breaks down a lot when the view is too narrow:

<img width="643" alt="plugins-manage-3-columns-broken" src="https://user-images.githubusercontent.com/664258/29774965-beac64f8-8c03-11e7-839f-b95035e7006c.png">

The Plugin Management page has two features that @iamtakashi's design proposal does not address yet.

First: when there are updates available for one or more plugins, the updatable ones are displayed first, and form a section that is separated by a vertical spacer from the rest:

<img width="740" alt="plugins-manage-available-upgrades" src="https://user-images.githubusercontent.com/664258/29775079-22a22aec-8c04-11e7-899d-5a04e5f8934b.png">

How will we implement this in the 3-column layout? A section with "Updatable plugins", followed by the rest?

Second: every plugin in the list can have a list of notices atached to it:

<img width="641" alt="plugins-manage-notices" src="https://user-images.githubusercontent.com/664258/29775194-708cc4ec-8c04-11e7-9355-dc3b24755a14.png">

This works great when there is just one column, but where are we going to put the notices in the much more crowded 3-column view?

Cc @iamtakashi @sirbrillig

sirbrillig commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -256,26 +251,55 @@ const PluginsBrowser = React.createClass( {
+		}
+
+		const navigation = this.props.category ? this.getNavigationBar() : this.getSearchBar();
+		const manageButton = this.props.isJetpackSite && this.getManageButton();
```
Why do we show this button in the sidebar for every site, but only show it here for Jetpack sites? I think we should do one or the other.

sirbrillig commented:
Why do we show the "Manage" button in the sidebar for every site, but only show it in in the browse page header for Jetpack sites? I think we should do one or the other. Probably we should show it for every site, but some of the docs imply otherwise (this is probably a design question for @iamtakashi.)

The grid looks great @jsnajdr! I don't have a good answer for where to put the notices, though. Maybe hide them under an alert icon that shows the notices full-screen when tapped? That could be a lot of work to add. We can discuss it tomorrow hopefully.

I changed the list in this PR description to checkboxes so we can see what still needs to be done (this is rough, of course, since there are many sub-tasks).

iamtakashi commented:
> Why do we show the "Manage" button in the sidebar for every site, but only show it in in the browse page header for Jetpack sites?

Yes, the "Manage" button in the sidebar is supposed to appear in only Jetpack site.

> For example, the layout breaks down a lot when the view is too narrow:

How about switching to 1-column layout when the viewport below 1040px wide?

> First: when there are updates available for one or more plugins, the updatable ones are displayed first, and form a section that is separated by a vertical spacer from the rest:

As per Slack conversation @jsnajdr and I had, let's show updatable ones first as it's been and remove the separation.

> Second: every plugin in the list can have a list of notices atached to it:

Good point. Let me think about it :)

iamtakashi commented:
> How about switching to 1-column layout when the viewport below 1040px wide?

I've just tested the branch. Although changing the breakpoint didn't help much, I still think it's ok to change it 1 column below 1040px. And adding `flex-wrap: wrap` to `.plugin-item__actions`, and having some spacing between them should make the look better. 

<img width="646" alt="screen shot 2017-08-29 at 14 31 22" src="https://user-images.githubusercontent.com/908665/29823511-ce172900-8cc6-11e7-8274-0e500464ea5e.png">

jsnajdr commented:
> Yes, the "Manage" button in the sidebar is supposed to appear in only Jetpack site.

@sirbrillig @iamtakashi That's a mistake on my side, I'll remove the "Manage" button for Simple sites.

Currently, Simple sites show a page with WordPress.com features shown as "fake" Jetpack plugins when I go to `/plugins/manage`. Will that page still exist after we're done with this project? And how do I navigate to it when there is no "Manage" button? Is the content going to be fully merged into the "Plugin Browser" view?

I'm pretty sure I've seen the answer already in one of the P2 or in the comments, I just can't find it now.

jsnajdr commented:
> How about switching to 1-column layout when the viewport below 1040px wide?

I just pushed a commit that moves the breakpoint to 1040px.

A plugin in the 1-column view now looks like this:

<img width="669" alt="plugin-1-column-new" src="https://user-images.githubusercontent.com/664258/29827601-b17f5e02-8cda-11e7-9489-e4af3cc00004.png">

It's a layout optimized for a narrow cell in the 3-column grid. Just used without change in one wide column.

In the original 1-column layout, a plugin looked like this:

<img width="670" alt="plugin-1-column-old" src="https://user-images.githubusercontent.com/664258/29827679-e953bb98-8cda-11e7-9456-b31d5bc733a7.png">

What do you think @iamtakashi ? Maybe we should keep the old layout for the 1-column view without any change, and use the new one only in 3-column view?

jsnajdr commented:
> Yes, the "Manage" button in the sidebar is supposed to appear in only Jetpack site.

Fixed! :tada:

iamtakashi commented:
> Maybe we should keep the old layout for the 1-column view without any change, and use the new one only in 3-column view?

If that wouldn't create a lot of more work, I think it's better as reduce the height.

jsnajdr commented:
> And adding flex-wrap: wrap to .plugin-item__actions, and having some spacing between them should make the look better.

Just implemented that in the latest push. It's a bit dissappointing that some actions are broken, while other remain on the same line:

<img width="811" alt="plugin-actions-wrap" src="https://user-images.githubusercontent.com/664258/29838044-49ad0e94-8cfa-11e7-9302-5b0f4a1025a0.png">

Also, wrapping of the upgrade notices is not very nice. The 3-column view displays everything without breaking only when the browser window is 1300px wide and more. I'm starting to think that 3 columns are maybe too many for the plugin management view.

> If that wouldn't create a lot of more work, I think it's better as reduce the height.

Yes, that should be a CSS-only change -- a few font sizes and flex directions will be different when a breakpoint is active, and the same markup can display both views equally well.

iamtakashi commented:
> I'm starting to think that 3 columns are maybe too many for the plugin management view.

I anticipated we need some design tweaks since everything was designed for 1 column, but I'd like to try for a bit to push a consistency in both views. I might give up if it becomes too much trouble though 😜

iamtakashi commented:
> Yes, that should be a CSS-only change -- a few font sizes and flex directions will be different when a breakpoint is active, and the same markup can display both views equally well.

OK, I'll work on the css for it.

a8ci18n commented on client/my-sites/plugins/main.jsx:
```diff
@@ -373,57 +378,65 @@ const PluginsMain = React.createClass( {
+						className="plugins__header-button"
+						aria-label={ this.props.translate( 'Browse all plugins', { context: 'button label' } ) }>
+						<Gridicon key="plus-icon" icon="plus-small" size={ 18 } />
+						<span className="plugins__header-button-text">{ this.props.translate( 'Add Plugin' ) }</span>
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Add a Plugin' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

jsnajdr commented on client/my-sites/plugins/plugin-action/style.scss:
```diff
@@ -33,26 +39,37 @@
-	text-transform: uppercase;
 	cursor: pointer;
 
+	@include breakpoint( '>1040px' ) {
```
@iamtakashi Changing the styles of the `plugin-action` and `plugin-activate-toggle` this way breaks layout of other places where these toggles are used.

First of them is the `PluginMeta` component. Have a Jetpack site selected and go to `plugins/jetpack`. The page has the toggles in the header area:

![plugin-meta](https://user-images.githubusercontent.com/664258/29861225-d99db102-8d68-11e7-9368-6301f68dc371.gif)

Second place is the `PluginSite` component. Select "All My Sites" in the site selector and to to `/plugins/amp`. At the bottom, there is a section for each site that has the plugin installed:

![plugin-site](https://user-images.githubusercontent.com/664258/29861304-2268f5b8-8d69-11e7-8308-81e8484d4bb3.gif)

That's why I was adding the style overrides for `.plugin-action` et al. as a nested selector in `plugin-item/style.scss`:

```css
.plugin-item {
  .plugin-action {
    ...
  }
}
```

iamtakashi replied:
Oh, I see! I'll fix it as well as the plugin icon you mentioned below. Sorry about that.

jsnajdr commented on client/my-sites/plugins/plugin-icon/style.scss:
```diff
@@ -4,20 +4,12 @@
-	height: 56px;
+	margin-right: 12px;
+	width: 48px;
+	height: 48px;
```
@iamtakashi The `PluginIcon` component is used at many places and at some of them, the style changes are not an improvement and most likely unintentional. The `/plugins/akismet` page looks like this in `master`:

<img width="501" alt="plugin-meta-production" src="https://user-images.githubusercontent.com/664258/29862167-12cd0e66-8d6c-11e7-8168-d8dd9215e71f.png">

The icon is large (56x56px) and has the same spacing (24px) on all 4 sides.

This is the new look in our branch:

<img width="502" alt="plugin-meta-branch" src="https://user-images.githubusercontent.com/664258/29862203-31fc47fc-8d6c-11e7-9ca2-bc2e26b5843a.png">

Small icon, uneven spacing. If your intent was to improve the styling only for `.plugin-item` and `.plugins-browser-item`, let's scope the CSS rules accordingly.

iamtakashi replied:
My bad. I'll fix this also.

a8ci18n commented on client/my-sites/plugins/main.jsx:
```diff
@@ -373,57 +378,66 @@ const PluginsMain = React.createClass( {
+		const addPluginButton = (
+			<PluginHeaderButton
+				icon="plus-small"
+				label={ this.props.translate( 'Add Plugin' ) }
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Add a Plugin' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +202,113 @@ const PluginsBrowser = React.createClass( {
+	getSearchBar() {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
+			this.props.translate( 'Security', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **34** times:
`translate( 'Security', { context: 'settings screen'} )` *ES Score: 17*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/plugins-browser/index.jsx:
```diff
@@ -195,76 +202,113 @@ const PluginsBrowser = React.createClass( {
 
+	getSearchBar() {
+		const suggestedSearches = [
+			this.props.translate( 'Engagement', { context: 'Plugins suggested search term' } ),
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Engagement' )` *ES Score: 26*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/update/add-plugins-page-first)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/my-sites/plugins/main.jsx:
```diff
@@ -373,57 +383,67 @@ const PluginsMain = React.createClass( {
+		const addPluginButton = (
+			<PluginHeaderButton
+				icon="plus"
+				label={ this.props.translate( 'Add Plugin' ) }
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Add a Plugin' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

sirbrillig commented:
Per a Slack discussion with @iamtakashi I removed the error messages shown per plugin since it seems that they were duplicated anyway in the global `NoticesList`.

That said, I'm not certain it's a safe change. A quick look seems to show that the errors included with the `PluginItem` carry quite a bit of meta-data and are passed with props in a clear way. The global `NoticesList` on the other hand seems to use a Flux store approach and the only data I could find was the message text itself.

The current `PluginItem` error notices were added 2 years ago in 2a7e7c91cd12010b62ba9c1b5fd8979543876755, but that really just changed how they were displayed. Prior to that, they were added before the open-sourcing of Calypso, in a commit entitled "Plugins: Integrate new stores." which seems to have been when the Plugins section was migrated to Flux. 

It seems likely to me that the inline errors approach might be outdated since we have global errors (did we, back then?), even if they are well defined.

iamtakashi commented:
Minor and it's possibly WIP, but I've noticed the "All" is no longer highlighted on `/plugins/`.

<img width="795" alt="screen shot 2017-08-31 at 00 54 46" src="https://user-images.githubusercontent.com/908665/29900081-3f06927e-8de7-11e7-9a79-7279520b41e8.png">

jsnajdr commented:
> It seems likely to me that the inline errors approach might be outdated since we have global errors (did we, back then?). On the other hand, if bulk actions (eg: bulk update) causes multiple errors, could they be lost or confusing if the errors are not paired with the plugin?

I spent some time today testing this and reading the code. The notices that are displayed inline with the plugin item and the ones displayed globally are identical -- the text, the buttons and other properties are the same.

The difference is that there is always only one global notice displayed -- the last one. It acts as a log (the object managing the data is actually called `LogStore`). That works great when activating/deactivating just one plugin.

But when working with multiple plugins and for some of them the action fails and succeeds for others, you don't get to see all the errors as global notices -- only a more or less complete summary. There is quite a complicated logic that creates messages like this:

<img width="672" alt="bulk-activate-notice" src="https://user-images.githubusercontent.com/664258/29929657-5f6d27a0-8e6c-11e7-8e28-b04ba82a8c51.png">

Some of them name the plugin explicitly, other just offer a summary like "2 plugins failed and 3 succeeded".

I see two options:
1. Accept the fact that when doing complicated bulk-editing and there is only partial success, you don't get to see all the errors.

2. Show the error inline like this:
<img width="321" alt="plugin-inline-notice" src="https://user-images.githubusercontent.com/664258/29930092-b6f5c2ba-8e6d-11e7-9a0b-fbd45b80dd00.png">
And show the full global notice after clicking.

jsnajdr commented:
I also fixed another edge case today: disabled plugins. That happens when Jetpack Manage is disabled on all Jetpack sites where the plugin is active. The plugin component has a `hasAllNoManageSites` prop in that case. It's shown as disabled and clicking on it displays a notice:
<img width="641" alt="plugins-manage-notices" src="https://user-images.githubusercontent.com/664258/29930253-3440a712-8e6e-11e7-88d0-5730151f98b7.png">

I changed this to a global notice on click.

The fact that all the inline notices are gone allows us to simplify the `PluginItem` component a lot -- see my last few commits today to see what I managed to remove :)

jsnajdr commented:
> Minor and it's possibly WIP, but I've noticed the "All" is no longer highlighted on /plugins/.

This was intentional thing that @sirbrillig did. Note that none of the other tabs ("Engagement", "Writing", ...) are ever displayed as selected. Clicking on them opens a pre-defined search:

![predefined-search](https://user-images.githubusercontent.com/664258/29930580-fd321f3e-8e6e-11e7-8b2d-d0258c75c352.gif)

iamtakashi commented:
> Note that none of the other tabs ("Engagement", "Writing", ...) are ever displayed as selected. Clicking on them opens a pre-defined search:

I know. I was saying the "All" tab specifically. But if I think about it more, we probably should remove the "All" anyway since we never show all the plugins there.

iamtakashi commented:
> I see two options:

Can we show a global notice for each plugin? I guess not, hence it has the name :) But I'm just curious the possibility. Otherwise, I prefer the second but I didn't quite get this bit.

> And show the full global notice after clicking.

After clicking... what?

enejb commented:
> It seems likely to me that the inline errors approach might be outdated since we have global errors (did we, back then?). On the other hand, if bulk actions (eg: bulk update) causes multiple errors, could they be lost or confusing if the errors are not paired with the plugin?

@sirbrillig to answer some of your questions here. 
As If I remember we did have the global notices but we didn't want the errors to be lost if things went wrong. Since the global notices get overwritten.

sirbrillig commented:
@iamtakashi I think he means "click on the 'error:' notice inside the card to see the full list of errors".

@enejb thanks for the perspective! I think "click to see the full notice" might solve the issue of notices getting overwritten.

sirbrillig commented:
> I know. I was saying the "All" tab specifically. But if I think about it more, we probably should remove the "All" anyway since we never show all the plugins there.

Agreed. It doesn't make much sense in this context. I removed it just now.

jsnajdr commented:
@iamtakashi Yes, I mean clicking on the small red notice inside the plugin item box to see the full one. Either as a popover, or just another global notice in the top-right corner.

Displaying all the notices as a sequence of global ones, as @enejb suggess, should work, too. But I like the clickable mini-notices idea more.

sirbrillig commented:
Found a bug with our search field hack. To reproduce:

1. On the Browse page, press the search icon or click a suggested search.
2. Then try typing into the search field. Pause a moment while the debounced search runs. Then type some more.

You'll notice that the search field loses focus when the debounced search runs, so if you type too slowly you won't be able to type at all.

jsnajdr commented:
> Found a bug with our search field hack. To reproduce:

Great catch @sirbrillig ! I know about this mistake for a few days, but I forgot to fix it :(

Go to the `handleSuggestedSearch` and move the `reinitializeSearch` into the returned arrow handler.

Right now, it's called several times on every render, which is of course very bad. It should be called only when suggested search link is actually clicked.

When the `WrappedSearch` is being reinitialized, the original component is removed from the DOM and new one is added. Focus on the old element is lost.

iamtakashi commented:
@sirbrillig @jsnajdr Thanks for clarifying. My only concern is that if the customers will realise that they need to click the inline notice to see more information. I guess we can add something like "Click here for more info.".

sirbrillig commented:
We're in good shape!

I Rebased to squash some commits and reviewed the checklist of tasks. I also added a task for the clickable mini-notices, although that could end up being outside the scope of this project. We need to decide tomorrow if we will finish [the remaining items](https://github.com/Automattic/wp-calypso/pull/17537#issue-252948511) or wait. cc @rralian for your thoughts. 

I initially was hoping to split this PR into multiple sequential PRs, but the nature of our changes makes that really difficult, particularly because the CSS changes span multiple components. I think we have two options:

1. Try to get someone to review this rather large PR. 
2. Try to extract out any changes that could become their own PRs (I've started work on this, but it's slow going).

I think the sensible PR splits are: 
- change routes around
- making a 3-column layout
- adding suggested search to Browse
- adding buttons to the header (with `PluginHeaderButton`)

Until we make a choice, I'll continue work on 2 tomorrow.

jsnajdr commented:
> Thanks for clarifying. My only concern is that if the customers will realise that they need to click the inline notice to see more information. I guess we can add something like "Click here for more info.".

@iamtakashi I prototyped two ideas how to improve on this.

First one is a compact notice with a "more" link:
<img width="351" alt="plugin-error-notice" src="https://user-images.githubusercontent.com/664258/29962619-18eed984-8f04-11e7-956d-6efd5b2ade62.png">

Second one is just a red warning icon in the corner. I think it kind of invites the user to hover or click on it to learn what it has to say:
<img width="351" alt="plugin-error-icon" src="https://user-images.githubusercontent.com/664258/29962671-424d1034-8f04-11e7-9e2c-74ba39a7dc79.png">

At the same time, a global notice saying something like "A lot of things have failed" is shown. That's the context in which the plugin-specific error UI is displayed.

iamtakashi commented:
@jsnajdr Thanks for prototyping those! Some feedback:

First one:
* "INFO" would probably better than "MORE".

Second one:
* I like the subtleness, but at the same time I'm worry that it can be missed easily. And how can I dismiss it without `X`?

It's a tough choice but I'm leaning towards the first one with "INFO" just because of the clarity.

iamtakashi commented:
I've noticed duplicated "Current Section" headers on the manage page with a small viewport.

<img width="412" alt="screen shot 2017-09-01 at 11 49 28" src="https://user-images.githubusercontent.com/908665/29966909-e4915380-8f0b-11e7-933a-543a13f1fb26.png">

jsnajdr commented:
> I've noticed duplicated "Current Section" headers on the manage page with a small viewport.

@iamtakashi That's a nice catch :) The bug is actually also in the `master` production version, not only in our branch:

<img width="514" alt="plugins-double-main" src="https://user-images.githubusercontent.com/664258/29973129-c27b07e6-8f2e-11e7-954d-153a046a5d61.png">

Steps to reproduce: go to the Plugins Management page for your Jetpack/Atomic site and search for something that's not found in your installed plugins. A "suggested" section with WP.org plugins will appear instead of the empty search results.

The root cause is that when the plugin browser is embedded in the plugin manager view, there are two `Main` components nested inside each other -- with two sidebars etc.

To fix this, we'll need to reorganize the components in `my-sites/plugins/plugins-browser` a bit. Let's do it in an independent PR.

sirbrillig commented:
> To fix this, we'll need to reorganize the components in my-sites/plugins/plugins-browser a bit. Let's do it in an independent PR.

@jsnajdr could you make an Issue for this, just in case we lose track?

sirbrillig commented:
Splitting into separate PRs:

- https://github.com/Automattic/wp-calypso/pull/17700
- https://github.com/Automattic/wp-calypso/pull/17702
- https://github.com/Automattic/wp-calypso/pull/17707
- ...

sirbrillig commented:
Rebased on top of #17719 until that gets merged

a8ci18n commented on client/my-sites/plugins/main.jsx:
```diff
@@ -344,6 +367,23 @@ const PluginsMain = React.createClass( {
+		return (
+			<HeaderButton
+				icon="plus"
+				label={ this.props.translate( 'Add Plugin' ) }
```
Hi! I've found a possible matching string that has already been translated **19** times:
`translate( 'Add a Plugin' )` *ES Score: 12*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

sirbrillig commented:
#17763 should be the last of things. 

Remaining nice-to-have tasks for this project, which may end up elsewhere:

- [ ]  Redirect /plugins/browse to /plugins, keeping filter, site, and search
- [ ] Add Jetpack features as "plugin" blocks in browse page
- [ ]  Add error message notice inside plugin item; click error notice to show error
- [ ]  Move "Upload Plugin" button in Manage page to header

sirbrillig commented:
Closing this PR for now, since all its code is elsewhere.
